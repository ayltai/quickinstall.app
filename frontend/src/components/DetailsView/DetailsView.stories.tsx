import type { Meta, StoryObj, } from '@storybook/react';

import { DetailsView, } from './DetailsView';

const meta : Meta<typeof DetailsView> = {
    component : DetailsView,
    title     : 'Components/Data Presentations/Details View',
    argTypes  : {
        description : {
            description : 'The description of the component',
        },
        infoUrl     : {
            description : 'The info URL of the component',
        },
        source      : {
            description : 'The source of the information',
        },
    },
};

export default meta;

type Story = StoryObj<typeof DetailsView>;

export const Default : Story = {
    args : {
        description : 'The description of the component',
        infoUrl     : 'https://example.com',
        source      : 'Homebrew',
    },
};
