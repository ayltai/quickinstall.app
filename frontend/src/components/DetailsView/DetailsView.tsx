import Box from '@mui/material/Box';
import Link from '@mui/material/Link';
import Stack from '@mui/material/Stack';
import Typography from '@mui/material/Typography';
import React, { type FC, } from 'react';
import { useTranslation, } from 'react-i18next';

import type { DetailsViewProps, } from './DetailsView.types';

/**
 * A details view is a container for displaying a description and optionally a URL for additional information and other arbitrary content.
 * @param color The color of the component
 * @param description The description of the component
 * @param infoUrl The URL of the information
 * @param source The source of the information
 * @param props Other props
 */
export const DetailsView : FC<DetailsViewProps> = ({
    color = 'primary',
    description,
    infoUrl,
    source,
    ...props
}) => {
    const { children, ...rest } = props;

    const { t, } = useTranslation();

    return (
        <Stack
            spacing={2}
            {...rest}>
            <Box>
                <Typography
                    variant='overline'
                    fontWeight='bold'>
                    {t('common.description')}
                </Typography>
                <Typography>
                    {description}
                </Typography>
            </Box>
            {infoUrl && (
                <Box>
                    <Typography
                        variant='overline'
                        fontWeight='bold'>
                        {t('common.website')}
                    </Typography>
                    <Link
                        color={color === 'info' ? 'info.main' : color}
                        href={infoUrl}
                        target='_blank'>
                        <Typography>
                            {infoUrl}
                        </Typography>
                    </Link>
                </Box>
            )}
            <Box>
                <Typography
                    variant='overline'
                    fontWeight='bold'>
                    {t('common.source')}
                </Typography>
                <Typography>
                    {source}
                </Typography>
            </Box>
            {children}
        </Stack>
    );
};
