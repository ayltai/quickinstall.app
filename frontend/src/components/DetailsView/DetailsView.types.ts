import type { HTMLAttributes, } from 'react';

export type DetailsViewProps = Omit<HTMLAttributes<HTMLDivElement>, 'color' | 'onChange'> & {
    color?       : 'info' | 'primary' | 'secondary',
    description? : string,
    infoUrl?     : string,
    source       : string,
};
