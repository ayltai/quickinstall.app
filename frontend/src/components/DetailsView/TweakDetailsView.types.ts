import type { DetailsViewProps, } from './DetailsView.types';

export type TweakDetailsViewProps = Omit<DetailsViewProps, 'children'> & {
    parameters : string[],
    types      : ('boolean' | 'number' | 'string')[],
    values     : (string | number | boolean)[],
    onChange?  : (parameter : string, value? : string | number | boolean) => void,
};
