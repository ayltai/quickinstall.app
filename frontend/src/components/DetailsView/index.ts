export { DetailsView, } from './DetailsView';
export type { DetailsViewProps, } from './DetailsView.types';
export { TweakDetailsView, } from './TweakDetailsView';
export type { TweakDetailsViewProps, } from './TweakDetailsView.types';
