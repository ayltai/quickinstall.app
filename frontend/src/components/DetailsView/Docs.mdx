import { Controls, Description, Meta, Title, Primary, Stories, } from '@storybook/blocks';

import * as ComponentStories from './DetailsView.stories';

<Meta of={ComponentStories} />

<Title />

<Description />

# How to use details views

Use details views,
* Don't use them in cards

# Default usage

<Primary />

# Parameters

<Controls sort='requiredFirst' />

<Stories
    title='Variants of buttons'
    includePrimary={false} />
