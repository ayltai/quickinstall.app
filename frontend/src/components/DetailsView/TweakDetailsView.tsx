import Divider from '@mui/material/Divider';
import Stack from '@mui/material/Stack';
import React, { type ComponentType, JSX, } from 'react';

import { DynamicTypedInput, } from '../DynamicTypedInput';
import { DetailsView, } from './DetailsView';
import type { DetailsViewProps, } from './DetailsView.types';
import type { TweakDetailsViewProps, } from './TweakDetailsView.types';

const withDynamicTypedInputs = (Component : ComponentType<DetailsViewProps>) => {
    const ComponentWithDynamicTypedInputs : (props : TweakDetailsViewProps) => JSX.Element = (props : TweakDetailsViewProps) => {
        const { parameters, types, values, onChange, ...rest } = props;

        return (
            <Component
                {...rest}
                color='secondary'>
                <Divider />
                <Stack
                    width='100%'
                    spacing={2}>
                    {parameters.map((parameter, index) => {
                        const handleChange = (value? : string | number | boolean) => onChange && onChange(parameter, value);

                        return (
                            <DynamicTypedInput
                                key={parameter}
                                color='secondary'
                                title={parameter}
                                type={types[index]}
                                value={values[index]}
                                onChange={handleChange} />
                        );
                    })}
                </Stack>
            </Component>
        );
    };

    return ComponentWithDynamicTypedInputs;
};

export const TweakDetailsView = withDynamicTypedInputs(DetailsView);
