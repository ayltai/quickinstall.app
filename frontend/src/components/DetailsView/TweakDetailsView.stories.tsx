import type { Meta, StoryObj, } from '@storybook/react';

import { TweakDetailsView, } from './TweakDetailsView';

const meta : Meta<typeof TweakDetailsView> = {
    component : TweakDetailsView,
    title     : 'Components/Data Presentations/Tweak Details View',
    tags      : [
        'autodocs',
    ],
    argTypes  : {
        description : {
            description : 'The description of the component',
        },
        infoUrl     : {
            description : 'The info URL of the component',
        },
        parameters  : {
            description : 'The parameters of the component',
        },
        types       : {
            description : 'The types of the component',
        },
        values      : {
            description : 'The values of the component',
        },
        onChange    : {
            description : 'The onChange of the component',
        },
    },
};

export default meta;

type Story = StoryObj<typeof TweakDetailsView>;

export const Default : Story = {
    args : {
        description : 'The description of the component',
        infoUrl     : 'https://example.com',
        parameters  : [
            'Width',
            'Height',
            'Disabled',
            'Colour',
        ],
        types       : [
            'number',
            'number',
            'boolean',
            'string',
        ],
        values      : [
            100,
            200,
            false,
            'red',
        ],
    },
};
