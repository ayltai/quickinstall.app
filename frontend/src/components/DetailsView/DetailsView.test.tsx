import React from 'react';

import { render, } from '../../utils/test';
import { DetailsView, } from './DetailsView';

describe('<DetailsView />', () => {
    it('renders correctly', () => expect(render(
        <DetailsView
            description='Description'
            infoUrl='https://example.com'
            source='Homebrew' />
    )).toMatchSnapshot());
});
