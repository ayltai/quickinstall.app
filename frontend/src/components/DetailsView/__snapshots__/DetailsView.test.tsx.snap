// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<DetailsView /> > renders correctly 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="MuiStack-root css-1sazv7p-MuiStack-root"
      >
        <div
          class="MuiBox-root css-0"
        >
          <span
            class="MuiTypography-root MuiTypography-overline css-1ooch7a-MuiTypography-root"
          >
            common.description
          </span>
          <p
            class="MuiTypography-root MuiTypography-body1 css-1che66z-MuiTypography-root"
          >
            Description
          </p>
        </div>
        <div
          class="MuiBox-root css-0"
        >
          <span
            class="MuiTypography-root MuiTypography-overline css-1ooch7a-MuiTypography-root"
          >
            common.website
          </span>
          <a
            class="MuiTypography-root MuiTypography-inherit MuiLink-root MuiLink-underlineAlways css-1b3q4o2-MuiTypography-root-MuiLink-root"
            href="https://example.com"
            target="_blank"
          >
            <p
              class="MuiTypography-root MuiTypography-body1 css-1che66z-MuiTypography-root"
            >
              https://example.com
            </p>
          </a>
        </div>
        <div
          class="MuiBox-root css-0"
        >
          <span
            class="MuiTypography-root MuiTypography-overline css-1ooch7a-MuiTypography-root"
          >
            common.source
          </span>
          <p
            class="MuiTypography-root MuiTypography-body1 css-1che66z-MuiTypography-root"
          >
            Homebrew
          </p>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="MuiStack-root css-1sazv7p-MuiStack-root"
    >
      <div
        class="MuiBox-root css-0"
      >
        <span
          class="MuiTypography-root MuiTypography-overline css-1ooch7a-MuiTypography-root"
        >
          common.description
        </span>
        <p
          class="MuiTypography-root MuiTypography-body1 css-1che66z-MuiTypography-root"
        >
          Description
        </p>
      </div>
      <div
        class="MuiBox-root css-0"
      >
        <span
          class="MuiTypography-root MuiTypography-overline css-1ooch7a-MuiTypography-root"
        >
          common.website
        </span>
        <a
          class="MuiTypography-root MuiTypography-inherit MuiLink-root MuiLink-underlineAlways css-1b3q4o2-MuiTypography-root-MuiLink-root"
          href="https://example.com"
          target="_blank"
        >
          <p
            class="MuiTypography-root MuiTypography-body1 css-1che66z-MuiTypography-root"
          >
            https://example.com
          </p>
        </a>
      </div>
      <div
        class="MuiBox-root css-0"
      >
        <span
          class="MuiTypography-root MuiTypography-overline css-1ooch7a-MuiTypography-root"
        >
          common.source
        </span>
        <p
          class="MuiTypography-root MuiTypography-body1 css-1che66z-MuiTypography-root"
        >
          Homebrew
        </p>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
